File: client/src/main/scala/stasis/client/encryption/Encoder.scala
package stasis.client.encryption

import org.apache.pekko.NotUsed
import org.apache.pekko.stream.scaladsl.Flow
import org.apache.pekko.util.ByteString

import stasis.client.encryption.secrets.DeviceFileSecret
import stasis.client.encryption.secrets.DeviceMetadataSecret

trait Encoder {
  def encrypt(fileSecret: DeviceFileSecret): Flow[ByteString, ByteString, NotUsed]
  def encrypt(metadataSecret: DeviceMetadataSecret): Flow[ByteString, ByteString, NotUsed]
  def maxPlaintextSize: Long
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/encryption/Encoder.kt
package stasis.client_android.lib.encryption

import okio.Sink
import okio.Source
import stasis.client_android.lib.encryption.secrets.DeviceFileSecret
import stasis.client_android.lib.encryption.secrets.DeviceMetadataSecret

interface Encoder {
    fun encrypt(source: Source, fileSecret: DeviceFileSecret): Source
    fun encrypt(sink: Sink, fileSecret: DeviceFileSecret): Sink
    fun encrypt(source: Source, metadataSecret: DeviceMetadataSecret): Source
    val maxPlaintextSize: Long
}
